utils/server_auth.go

```diff
@@ -0,0 +1,342 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+package utils
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/director"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	TokenSource int
+	TokenIssuer int
+	AuthOption  struct {
+		Sources []TokenSource
+		Issuers []TokenIssuer
+		Scopes  []string
+	}
+	AuthChecker interface {
+		FederationCheck(ctx *gin.Context, token string, scopes []string) error
+		DirectorCheck(ctx *gin.Context, token string, scopes []string) error
+		IssuerCheck(ctx *gin.Context, token string, scopes []string) error
+	}
+	AuthCheckImpl struct{}
+)
+
+const (
+	Header TokenSource = iota // "Authorization" header
+	Cookie                    // "login" cookie
+	Authz                     // "authz" query parameter
+)
+
+const (
+	Federation TokenIssuer = iota
+	Director
+	Issuer
+)
+
+var (
+	federationJWK *jwk.Cache
+	authChecker   AuthChecker
+)
+
+func init() {
+	authChecker = &AuthCheckImpl{}
+}
+
+// Return if desiredScopes contains the tokenScope and it's case-insensitive
+func scopeContains(tokenScope string, desiredScopes []string) bool {
+	for _, sc := range desiredScopes {
+		if strings.EqualFold(sc, tokenScope) {
+			return true
+		}
+	}
+	return false
+}
+
+// Creates a validator that checks if a token's scope matches the given scope: matchScope.
+// Will pass the check if no "anyScopes".
+// Will pass the check if one token scope matches ANY item in "anyScopes"
+func createScopeValidator(anyScopes []string) jwt.ValidatorFunc {
+
+	return jwt.ValidatorFunc(func(_ context.Context, tok jwt.Token) jwt.ValidationError {
+		// If no scope is present, always return true
+		if len(anyScopes) == 0 {
+			return nil
+		}
+		scope_any, present := tok.Get("scope")
+		if !present {
+			return jwt.NewValidationError(errors.New("No scope is present; required for authorization"))
+		}
+		scope, ok := scope_any.(string)
+		if !ok {
+			return jwt.NewValidationError(errors.New("scope claim in token is not string-valued"))
+		}
+
+		for _, tokenScope := range strings.Split(scope, " ") {
+			// As long as there's one scope in the token that matches the pool of desriedScopes
+			// we say it's valid
+			if scopeContains(tokenScope, anyScopes) {
+				return nil
+			}
+		}
+		return jwt.NewValidationError(errors.New(fmt.Sprint("Token does not contain any of the scopes: ", anyScopes)))
+	})
+}
+
+// Checks that the given token was signed by the federation jwk and also checks that the token has the expected scope
```

Comment:
Not a request for change, but for clarification -- what is the `federation jwk`, and where does it live? If it lives at the metadata discovery endpoint returned by `param.Federation_DiscoveryUrl`, which is currently a static server, how would one actually get a token signed by the private key? Would that have to be hand-created?

Reply:
Yes! We don't have any token created by federation now and this function will soon be merged with DirectorCheck because director will soon become a federation itself in #417 

With this being said, It seems safe to remove `FederationCheck` from the code if it's never being used. However, I would ask for @turetske first to see if there's a reason to keep it as I vaguely remember Emma created this function originally as part of the token check process for Prometheus query engine endpoint.

Reply:
I believe the idea was from Brian, basically if someone has a token signed by the federation key then they have access to the prometheus metrics. It would have to be hand created. I believe it was mostly there for the future.